### ISSUE TITLE
Expressing Multiplicity of same type of Measurement using CoRIM Schema

### ISSUE THREAD

**Issue by yogeshbdeshpande** [ietf-rats-wg/draft-ietf-rats-corim#256]
With recent discussions with various teams using CoRIMs with their own profile it has become evident that there is a requirement to express in CoRIM, the following:

_For a given single Environment, there can be multiple measurements of the same type, i.e. For an Environment there can be a series of SVN’s pertaining to the environment. The other example is: for a given environment, it can have multiple acceptable states where each state has a unique measurement expressed either by a Digest or a Serial Number or a UUID (each of these are entries in a Measurement Values Map), in the base CoRIM schema._

This issue tracks the requirement and CoRIM Authors are requested to provide suitable possible options to debate on. 

Here are few which was discussed during Extended Meeting 14th August 2024:

Option 1: Individual CoRIM Profile extends the Measurement Values Map. The singular Measurement is expressed using Base CoRIM Schema and the Extension slot offer mechanism for expressing multiplicity.

Pros: The base schema is NOT impacted. The Implementation is specific to required profile where individual measurements for an Environment is uniquely identified by the profile owner. The matching rules are profile specific

Cons: Multiple profiles requiring a similar extension will end up duplicating implementation. A single Verifier supporting these profiles will have difficulty scaling the implementation.

**Comment by yogeshbdeshpande** [comment]
Option 2: The other option is to express different measurements using multiplicity of Reference Value Triples, each having the same Environment identifier.

**Comment by deeglaze** [comment]
I would certainly give the guidance that if you use mkey, then the mval should have exactly one value. I wouldn't require it.

**Comment by thomas-fossati** [comment]
We need to be able to describe two multiplicity-related situations:

1. multiplicity of acceptable states for a given environment (essentially, multiplicity which describes alternatives);
2. multiplicity of measurements that represent, _en bloc_, one acceptable state of a given environment.

For the former, multiple reference value triples, in the same or different CoMIDs/CoRIMs, with the same environment (and mkey, sigh!) seem to be the right construct.

For the latter, I think a simple and clear way to model is to group all the measurements into the same measurement map using a map/array type.

**Comment by thomas-fossati** [comment]
Quick follow-up.  The problem with spreading (rather than bunching) in case 2, is that spreading creates an ambiguity about what exactly constitutes the wanted TCB image in case there are less acceptable states than the cartesian product.  E.g.: $s_1 = [ A, B ]$ and $s_2 = [ C, D ]$ are OK, but $s_3 = [ A, D ]$ and $s_4 = [ C, B ]$ are not.

**Comment by deeglaze** [comment]
Agreed with Thomas that my usage of mkey above is more `en bloc` and not multiple acceptable states. If you want multiple acceptable states irrespective of other pieces of information, then `mval` itself should have multiplicity. If you need different combinations of acceptable states, you should have different CoMIDs to express those.

**Comment by deeglaze** [comment]
Certainly at this point it sounds more like we need to better account for reference value components as data bindings that can be given more nuanced handling by a policy that references the values bound to same names.

**Comment by yogeshbdeshpande** [comment]
> We then say $measured-element-type-choice is explicitly local to the env

@deeglaze My comment on Option 2 was focused on the existing CoRIM Schema, where `mkey` exits in `measurement-map`.
In this model, one needs to repeatedly insert the same `environment identity` for `N` number of Measurements belonging to same environment, when using `N` triples!

**Comment by deeglaze** [comment]
> In this model, one needs to repeatedly insert the same environment identity for N number of Measurements belonging to same environment, when using N Ref Value triples!

Insert the same environment identity where? The ACS? I don't think you have to store copies of the same environment. Just a reference to one. If the scaling problem is continually needing to match the same environment map, there's nothing stopping a Verifier from representing that nesting behavior in a more optimal way.

**Comment by deeglaze** [comment]
The _en bloc_ interpretation of all mkeys in the multiple measurement-maps suggests that the `authorized-by` ought to be for the collection and not for each of the mkeys themselves. In that case I'd want to change the multiplicity to something else

```
reference-triple-record = [environment-map, authority-and-claim-values-map / keyed-claim-map]
authority-and-claim-values-map = {
  &(cval: 1) => claim-values-map
  &(authorized-by: 2) => [ + authority-type-choice ]
}
keyed-claim-map = {
  &(claims: 0) => [ + key-claim-record ]
  &(authorized-by: 2) => [ + authority-type-choice ]
}
key-claim-record = [ckey: uint, cval: claim-values-map]
```

**Comment by yogeshbdeshpande** [comment]
> > In this model, one needs to repeatedly insert the same environment identity for N number of Measurements belonging to same environment, when using N Ref Value triples!
> 
> Insert the same environment identity where? The ACS? I don't think you have to store copies of the same environment. Just a reference to one. If the scaling problem is continually needing to match the same environment map, there's nothing stopping a Verifier from representing that nesting behavior in a more optimal way.

@deeglaze I was referring to CoRIM that a supply chain actor needs to specify to the Verifier. I am not referring to ACS which is an Internal representation where of course smart filtering will remove such need anyway!

**Comment by deeglaze** [comment]
> @deeglaze I was referring to CoRIM that a supply chain actor needs to specify to the Verifier. I am not referring to ACS which is an Internal representation where of course smart filtering will remove such need anyway!

Ah you were just talking about multiple possible measurements in the same environment, not different measurements of the same type (e.g., TDX's 16 CPU SVNs) in the same environment. Understood.

**Comment by thomas-fossati** [comment]
> Agreed with Thomas that my usage of mkey above is more `en bloc` and not multiple acceptable states. If you want multiple acceptable states irrespective of other pieces of information, then `mval` itself should have multiplicity.

This requires changing a non-leaf element of the schema.  IMHO we should look for minimally intrusive modifications first, and do disruptive ones only if there are no good alternatives.

> If you need different combinations of acceptable states, you should have different CoMIDs to express those.

Or different triples within the same CoMID.  (A CoMID does not define any semantic boundary, it's just a packaging construct.)

**Comment by deeglaze** [comment]
> > Agreed with Thomas that my usage of mkey above is more `en bloc` and not multiple acceptable states. If you want multiple acceptable states irrespective of other pieces of information, then `mval` itself should have multiplicity.
> 
> This requires changing a non-leaf element of the schema. IMHO we should look for minimally intrusive modifications first, and do disruptive ones only if there are no good alternatives.
> 
Yes, fair enough. Do we want to add a generically typed tag, then, that is for `measurement-values-map / codepoint / T` where we can always have `#6.5xx([ + T ])` to represent "any of these are allowed irrespective of surrounding context"?

> > If you need different combinations of acceptable states, you should have different CoMIDs to express those.
> 
> Or different triples within the same CoMID. (A CoMID does not define any semantic boundary, it's just a packaging construct.)

Okay, agreed. I was still working from the idea that we needed to merge

**Comment by deeglaze** [comment]
We've decided that disjunction is at the triple level.
